File: crates/wg-python/tests/smoke.py
"""End-to-end smoke test for wg-python.

Run after `maturin develop` (or `pip install` of the wheel).
Exercises every public method against a temp DB so we catch breakage early.
"""

import os
import shutil
import tempfile

import wg_python as wg


def main() -> None:
    tmp = tempfile.mkdtemp(prefix="wg-py-smoke-")
    db = os.path.join(tmp, "test.redb")
    try:
        g = wg.WikiGraph(db)

        # Entity CRUD
        eid_redis = g.entity_add(
            "Redis",
            entity_type="technology",
            tags=["cache", "infra"],
            aliases=["redis-server"],
        )
        eid_postgres = g.entity_add("Postgres", entity_type="technology")
        assert g.resolve_entity("Redis") == eid_redis
        assert g.resolve_entity("redis-server") == eid_redis  # alias

        e = g.entity_get("Redis")
        assert e["name"] == "Redis"
        assert "cache" in e["tags"]

        ents = g.entity_list(limit=10)
        assert len(ents) == 2
        assert {x["name"] for x in ents} == {"Redis", "Postgres"}

        # Facts
        fid = g.fact_add(
            "Redis Sentinel provides high availability",
            entity_ids=[eid_redis],
            fact_type="decision",
            tags=["ha"],
            confidence=0.9,
        )
        fact = g.fact_get(fid)
        assert fact["content"].startswith("Redis Sentinel")

        facts = g.fact_list(entity="Redis", limit=10)
        assert len(facts) == 1

        # Batch insert — single redb write txn for the whole list.
        many_ids = g.fact_add_many([
            {"content": "Redis Cluster shards keys by hash slot",
             "entity_ids": [eid_redis], "fact_type": "pattern"},
            {"content": "Redis 7 introduces Functions and ACL improvements",
             "entity_ids": [eid_redis], "fact_type": "note", "confidence": 0.85},
            {"content": "Postgres logical replication is the default",
             "entity_ids": [eid_postgres], "fact_type": "convention"},
        ])
        assert len(many_ids) == 3
        assert all(isinstance(x, str) for x in many_ids)
        # The new facts are findable.
        for fid_ in many_ids:
            rec = g.fact_get(fid_)
            assert rec["id"] == fid_

        # Relations
        g.relation_add("Redis", "Postgres", "alternative_to")
        rels = g.relations_get("Redis", direction="forward")
        assert len(rels) == 1

        # Search (BM25 — no semantic model loaded for the smoke test, but the
        # call path should still succeed and return results from BM25 only).
        try:
            hits = g.search("high availability", limit=5)
            print(f"search hits: {len(hits)}")
        except RuntimeError as e:
            # If semantic model isn't downloaded, hybrid search may fail — that's
            # an environment issue, not a binding issue. Don't fail the test.
            print(f"search skipped: {e}")

        # Graph
        traverse = g.traverse("Redis", depth=1, direction="both")
        assert "entities" in traverse
        path = g.path_find("Redis", "Postgres")
        assert path is not None and len(path) >= 1

        # Lint / stats
        issues = g.lint()
        stats = g.stats()
        print(f"stats: {stats}")
        print(f"lint issues: {len(issues)}")

        # Query (unified) — relies on hybrid search; tolerate failure on that
        # but verify the call dispatches.
        try:
            ctx = g.query("Redis", limit=3, depth=1, recent_limit=3)
            assert ctx["topic"] == "Redis"
            assert ctx["entity"]["name"] == "Redis"
            print(f"query keys: {list(ctx.keys())}")
        except RuntimeError as e:
            print(f"query skipped: {e}")

        # Validity windows: supersede the first fact with a new one and verify
        # current_only filtering hides it.
        new_fid = g.fact_add(
            "Redis Sentinel + Cluster supersedes Sentinel-only HA",
            entity_ids=[eid_redis],
            fact_type="decision",
        )
        g.fact_supersede(fid, new_fid)
        old = g.fact_get(fid)
        assert old["superseded_at"] is not None, "fact_supersede should set superseded_at"
        assert old["superseded_by"] == new_fid

        all_facts = g.fact_list(entity="Redis")
        current_facts = g.fact_list(entity="Redis", current_only=True)
        assert len(current_facts) == len(all_facts) - 1, "current_only should hide superseded"

        # Cleanup writes
        g.fact_delete(fid)
        g.fact_delete(new_fid)
        g.relation_remove("Redis", "Postgres", "alternative_to")
        g.entity_delete("Postgres")

        print("OK: wg-python smoke test passed")
    finally:
        shutil.rmtree(tmp, ignore_errors=True)


if __name__ == "__main__":
    main()

File: crates/wg-napi/tests/smoke.js
// End-to-end smoke test for wg-napi.
//
// Run: `npm run build && node tests/smoke.js` from crates/wg-napi.

const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');

const { WgStore, version } = require('..');

function main() {
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'wg-napi-smoke-'));
  const db = path.join(tmp, 'test.redb');

  try {
    const g = new WgStore(db);
    console.log(`wg-napi version: ${version()}`);

    // Entity CRUD
    const eidRedis = g.entityAdd('Redis', {
      entityType: 'technology',
      tags: ['cache', 'infra'],
      aliases: ['redis-server'],
    });
    g.entityAdd('Postgres', { entityType: 'technology' });
    if (g.resolveEntity('Redis') !== eidRedis) throw new Error('resolve_entity mismatch');
    if (g.resolveEntity('redis-server') !== eidRedis) throw new Error('alias resolution failed');

    const e = JSON.parse(g.entityGet('Redis'));
    if (e.name !== 'Redis') throw new Error(`expected Redis, got ${e.name}`);
    if (!e.tags.includes('cache')) throw new Error('missing cache tag');

    const ents = JSON.parse(g.entityList({ limit: 10 }));
    if (ents.length !== 2) throw new Error(`expected 2 entities, got ${ents.length}`);

    // Facts
    const fid = g.factAdd('Redis Sentinel provides high availability', {
      entityIds: [eidRedis],
      factType: 'decision',
      tags: ['ha'],
      sourceId: 'alpha',
      confidence: 0.9,
    });
    const fact = JSON.parse(g.factGet(fid));
    if (!fact.content.startsWith('Redis Sentinel')) throw new Error('fact content mismatch');

    const facts = JSON.parse(g.factList({ entity: 'Redis', limit: 10, sourceId: 'alpha' }));
    if (facts.length !== 1) throw new Error(`expected 1 fact, got ${facts.length}`);

    // Batch insert — single redb write txn for the whole array.
    const manyIds = g.factAddMany([
      { content: 'Redis Cluster shards by hash slot', entityIds: [eidRedis], factType: 'pattern' },
      { content: 'Redis 7 introduces Functions and ACL improvements',
        entityIds: [eidRedis], factType: 'note', confidence: 0.85 },
      { content: 'Postgres logical replication is the default',
        entityIds: [g.resolveEntity('Postgres')], factType: 'convention' },
    ]);
    if (manyIds.length !== 3) throw new Error(`expected 3 batch ids, got ${manyIds.length}`);
    for (const id of manyIds) {
      const rec = JSON.parse(g.factGet(id));
      if (rec.id !== id) throw new Error(`fact_get round-trip failed for ${id}`);
    }

    // Relations
    g.relationAdd('Redis', 'Postgres', 'alternative_to');
    const rels = JSON.parse(g.relationsGet('Redis', 'forward'));
    if (rels.length !== 1) throw new Error(`expected 1 relation, got ${rels.length}`);

    // Search (BM25 + semantic — semantic may no-op without model, but must dispatch)
    try {
      const hits = JSON.parse(g.search('high availability', { limit: 5, bm25Only: true, sourceId: 'alpha' }));
      console.log(`search hits: ${hits.length}`);
    } catch (e) {
      console.log(`search skipped: ${e.message}`);
    }

    // Graph
    const traverse = JSON.parse(g.traverse('Redis', { depth: 1, direction: 'both' }));
    if (!Array.isArray(traverse.entities)) throw new Error('traverse.entities not array');
    const found = JSON.parse(g.pathFind('Redis', 'Postgres'));
    if (!found || found.length < 1) throw new Error('expected path Redis -> Postgres');

    // Lint / stats
    const issues = JSON.parse(g.lint());
    const stats = JSON.parse(g.stats());
    console.log(`stats:`, stats);
    console.log(`lint issues: ${issues.length}`);

    // Query (unified)
    try {
      const ctx = JSON.parse(g.query('Redis', { limit: 3, depth: 1, recentLimit: 3 }));
      if (ctx.topic !== 'Redis') throw new Error('query topic mismatch');
      if (ctx.entity?.name !== 'Redis') throw new Error('query entity mismatch');
      console.log(`query keys: [${Object.keys(ctx).join(', ')}]`);
    } catch (e) {
      console.log(`query skipped: ${e.message}`);
    }

    // Validity windows
    const newFid = g.factAdd('Redis Sentinel + Cluster supersedes Sentinel-only HA', {
      entityIds: [eidRedis],
      factType: 'decision',
    });
    g.factSupersede(fid, newFid);
    const oldFact = JSON.parse(g.factGet(fid));
    if (oldFact.superseded_at == null) throw new Error('expected superseded_at to be set');
    if (oldFact.superseded_by !== newFid) throw new Error('expected superseded_by to point to newFid');

    const allFacts = JSON.parse(g.factList({ entity: 'Redis' }));
    const currentFacts = JSON.parse(g.factList({ entity: 'Redis', currentOnly: true }));
    if (currentFacts.length !== allFacts.length - 1) {
      throw new Error(`expected currentOnly to hide 1 fact (all=${allFacts.length}, current=${currentFacts.length})`);
    }

    // Cleanup
    g.factDelete(fid);
    g.factDelete(newFid);
    g.relationRemove('Redis', 'Postgres', 'alternative_to');
    g.entityDelete('Postgres');

    console.log('OK: wg-napi smoke test passed');
  } finally {
    fs.rmSync(tmp, { recursive: true, force: true });
  }
}

main();
